src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Core.cs

```diff
@@ -2,45 +2,138 @@
+                "System.Linq",
+                "System.Linq.Expressions",
+                "System.Memory",
+                "System.Reflection.Metadata",
+                "System.Reflection.Primitives",
+                "System.Resources.ResourceManager",
+                "System.Runtime",
+                "System.Runtime.CompilerServices.Unsafe",
+                "System.Runtime.Extensions",
+                "System.Runtime.InteropServices",
+                "System.Runtime.Loader",
+                "System.Runtime.Numerics",
+                "System.Runtime.Serialization.Primitives",
+                "System.Security.Cryptography.Algorithms",
+                "System.Security.Cryptography.Primitives",
+                "System.Text.Encoding.CodePages",
+                "System.Text.Encoding.Extensions",
+                "System.Text.RegularExpressions",
+                "System.Threading",
+                "System.Threading.Tasks",
+                "System.Threading.Tasks.Parallel",
+                "System.Threading.Thread",
+                "System.Threading.ThreadPool",
+                "System.Xml.ReaderWriter",
+                "System.Xml.XDocument",
+                "System.Xml.XPath.XDocument");
+
+        private readonly object _guard = new object();
+        private readonly Dictionary<string, AssemblyLoadContext> _loadContextByDirectory = new Dictionary<string, AssemblyLoadContext>();
 
         protected override Assembly LoadFromPathImpl(string fullPath)
         {
-            //.NET Native doesn't support AssemblyLoadContext.GetLoadContext. 
-            // Initializing the _loadContext in the .ctor would cause
-            // .NET Native builds to fail because the .ctor is called. 
-            // However, LoadFromPathImpl is never called in .NET Native, so 
-            // we do a lazy initialization here to make .NET Native builds happy.
-            if (_loadContext == null)
+            AssemblyLoadContext? loadContext;
+
+            var fullDirectoryPath = Path.GetDirectoryName(fullPath) ?? throw new ArgumentException(message: null, paramName: nameof(fullPath));
+            lock (_guard)
             {
-                AssemblyLoadContext loadContext = AssemblyLoadContext.GetLoadContext(typeof(DefaultAnalyzerAssemblyLoader).GetTypeInfo().Assembly);
+                if (!_loadContextByDirectory.TryGetValue(fullDirectoryPath, out loadContext))
+                {
+                    loadContext = new DirectoryLoadContext(fullDirectoryPath, this);
+                    _loadContextByDirectory[fullDirectoryPath] = loadContext;
+                }
+            }
+
+            var name = AssemblyName.GetAssemblyName(fullPath);
+            return loadContext.LoadFromAssemblyName(name);
+        }
 
-                if (System.Threading.Interlocked.CompareExchange(ref _loadContext, loadContext, null) == null)
+        private bool ShouldLoadInAnalyzerContext(string? assemblySimpleName, string fullPath)
```

Comment:
Think we should restrict `assemblySimpleName` to non-nullable values here and prune out `null` values earlier. Basically if `AssemblyLoadContext.Load` gets called with a null `AssemblyName.Name` value then we should bail early. It's a busted scenario.

Reply:
I just added a `!`. I don't think it's likely to come up in real code, and this code is resilient to it without actually needing to go out of our way to null-check the simple name.